### PR TITLE
Fix 2 possible crashes for 1.8

### LINKED ISSUE #71: Failed to launch with OptiFine 1.8.8
org.spongepowered.asm.mixin.injection.throwables.InjectionError: LVT in net/minecraft/crash/CrashReport::func_71498_d()Ljava/lang/String; has incompatible changes at opcode 132 in callback mixin.mixinbooter.init.json:CrashReportMixin from mod unknown-owner->@Inject::afterStackTracePopulation(Lorg/spongepowered/asm/mixin/injection/callback/CallbackInfoReturnable;Ljava/io/StringWriter;Ljava/io/PrintWriter;Ljava/lang/Throwable;)V
Expected: [Ljava/io/StringWriter;, Ljava/io/PrintWriter;, Ljava/lang/Throwable;]
Found: [Ljava/io/StringWriter;, Ljava/io/PrintWriter;, Ljava/lang/Object;]
Available: [Ljava/io/StringWriter;, Ljava/io/PrintWriter;, Ljava/lang/Object;, Ljava/lang/String;, Ljava/lang/Throwable;]

https://mclo.gs/dMVwJVY

### PR DESCRIPTION
**Pull request by HowardZHY**
Closes #71

### CODE REVIEW

File: src/main/java/zone/rong/mixinbooter/MixinBooterMixinPlugin.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    public void onLoad(String mixinPackage) {}
+
+    @Override
+    public String getRefMapperConfig() {
+        return null;
+    }
+
+    @Override
+    public boolean shouldApplyMixin(String targetClassName, String mixinClassName) {
+        String version = Loader.instance().getMCVersionString();
```
Comment (Ecdcaeb) [inline src/main/java/zone/rong/mixinbooter/MixinBooterMixinPlugin.java:23]: `net.minecraftforge.fml.common.Loader` ?
Or, `(String)net.minecraftforge.fml.relauncher.data()[4]` ?

Comment (HowardZHY) [reply src/main/java/zone/rong/mixinbooter/MixinBooterMixinPlugin.java:23]: > `net.minecraftforge.fml.common.Loader.instance().getMCVersionString()` ? I would recommend using `(String)net.minecraftforge.fml.relauncher.FMLInjectionData.data()[4]` .

https://github.com/MinecraftForge/FML/blob/d4ded9d6e218ac097990e836676bbe22b47e5966/src/main/java/net/minecraftforge/fml/relauncher/FMLInjectionData.java

The 1.8 version getProperty has chances to be missing, that can be caused by bad launcher installation

Comment (Ecdcaeb) [reply src/main/java/zone/rong/mixinbooter/MixinBooterMixinPlugin.java:23]: > > `net.minecraftforge.fml.common.Loader.instance().getMCVersionString()` ? I would recommend using `(String)net.minecraftforge.fml.relauncher.FMLInjectionData.data()[4]` .
> 
> https://github.com/MinecraftForge/FML/blob/d4ded9d6e218ac097990e836676bbe22b47e5966/src/main/java/net/minecraftforge/fml/relauncher/FMLInjectionData.java
> 
> The 1.8 version getProperty has chances to be missing, that can be caused by bad launcher installation

Equal

https://github.com/MinecraftForge/FML/blob/d4ded9d6e218ac097990e836676bbe22b47e5966/src/main/java/net/minecraftforge/fml/common/launcher/FMLDeobfTweaker.java#L39

the data in `Loader` comes from `[FMLInjectionData](https://github.com/MinecraftForge/FML/blob/d4ded9d6e218ac097990e836676bbe22b47e5966/src/main/java/net/minecraftforge/fml/relauncher/FMLInjectionData.java)`

Comment (HowardZHY) [reply src/main/java/zone/rong/mixinbooter/MixinBooterMixinPlugin.java:23]: There's no reason to get data from an Object[] that might contain unexpected content.
Just use the getter that can always return the thing wanted

**Comment by Ecdcaeb** [comment]
Good enough for a stop-gap until mixinbooter is compatible with 1.8.

**Review by Ecdcaeb — APPROVED** [review]
Good enough for a stop-gap until mixinbooter is compatible with 1.8.

**Review by Rongmario — CHANGES_REQUESTED** [review]

File: src/main/java/zone/rong/mixinbooter/MixinBooterMixinPlugin.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+        return null;
+    }
+
+    @Override
+    public boolean shouldApplyMixin(String targetClassName, String mixinClassName) {
+        String version = (String) net.minecraftforge.fml.relauncher.FMLInjectionData.data()[4];
+        MixinBooterPlugin.LOGGER.info("Detected MC Version : " + version);
+        if (mixinClassName.contains("CrashReport")) {
+            // 1.8 & 1.8.8
+            if (version.endsWith(".8")) {
```
[the remark this answers is not in the document]
Comment (HowardZHY) [reply src/main/java/zone/rong/mixinbooter/MixinBooterMixinPlugin.java:26]: > Might be better to parse the version here and check if its `1.8 >= n <= 1.8.8`

Forge itself doesn't support 1.8.1-1.8.7, this is neat enought

File: src/main/java/zone/rong/mixinbooter/mixin/LoadControllerMixin.java [hunk 3]
```diff
@@ -49,7 +49,7 @@ private void beforeConstructing(LoaderState state, Object[] eventData, CallbackI
                     log = true;
                 }
                 Class<?> clazz = Class.forName(asmData.getClassName());
-                MixinBooterPlugin.LOGGER.info("Instantiating {} for its mixins.", clazz);
+                MixinBooterPlugin.LOGGER.info("Instantiating " + clazz + " for its mixins.");
```
[the remark this answers is not in the document]
Comment (HowardZHY) [reply src/main/java/zone/rong/mixinbooter/mixin/LoadControllerMixin.java:52]: > Why these changes?

This is the fix of the second java.lang.noSuchMethodError crash, caused by `info(String, Object)` doesn't exist in older log4j versions

[the remark this answers is not in the document]
Comment (Rongmario) [reply src/main/java/zone/rong/mixinbooter/mixin/LoadControllerMixin.java:52]: Interesting..

**Review by Rongmario — APPROVED** [review]

**Pull request merged**